geo/src/algorithm/mean_center.rs

```diff
@@ -0,0 +1,694 @@
+/// # use geo::{Point, MultiPoint, Coord};
+/// let coords: Vec<Point> = vec![
+///     Coord { x: 0.0, y: 0.0 }.into(),
+///     Coord { x: 1.0, y: 0.0 }.into(),
+///     Coord { x: 0.0, y: 1.0 }.into(),
+///     Coord { x: 1.0, y: 1.0 }.into(),
+/// ];
+///
+/// let mpnt = MultiPoint::new(coords);
+/// let mean_center = mpnt.mean_center();
+/// assert_eq!(mean_center, Point::new(0.5, 0.5));
+///
+/// let weighted_center = mpnt.weighted_mean_center(&[0.0, 1.0, 2.0, 2.0]);
+/// assert_eq!(weighted_center, Point::new(0.6, 0.8));
+/// ```
+
+pub trait MeanCenter {
+    type Output;
+    /// Return the unweighted mean center of a geometry.
+    fn mean_center(&self) -> Self::Output;
+
+    /// Return the weighted mean center of a geometry.
+    /// The weights are cycled if there are fewer weights than geometries.
+    fn weighted_mean_center(&self, weights: &[f64]) -> Self::Output;
+}
+
+macro_rules! impl_mean_center_point {
+    ($type:ty) => {
+        impl MeanCenter for $type {
+            type Output = Point;
+            fn mean_center(&self) -> Self::Output {
+                self.centroid()
+            }
+
+            fn weighted_mean_center(&self, weights: &[f64]) -> Self::Output {
+                let mut x_sum = 0.0;
+                let mut y_sum = 0.0;
+                let mut weight_sum = 0.0;
+
+                for (coord, weight) in self.coords_iter().zip(weights.iter().cycle()) {
+                    let (xi, yi) = coord.x_y();
+                    x_sum += xi * weight;
+                    y_sum += yi * weight;
+                    weight_sum += weight;
+                }
+
+                let x = x_sum / weight_sum;
+                let y = y_sum / weight_sum;
+
+                Point::new(x, y)
+            }
+        }
+    };
+}
+
+impl_mean_center_point!(Line);
+impl_mean_center_point!(Point);
+impl_mean_center_point!(Rect);
+impl_mean_center_point!(Triangle);
+
+macro_rules! impl_mean_center_option {
+    ($type:ty) => {
+        impl MeanCenter for $type {
+            type Output = Option<Point>;
+            fn mean_center(&self) -> Self::Output {
+                self.centroid()
+            }
+
+            fn weighted_mean_center(&self, weights: &[f64]) -> Self::Output {
+                let mut x_sum = 0.0;
+                let mut y_sum = 0.0;
+                let mut weight_sum = 0.0;
+                let n = self.coords_count();
+                if n == 0 {
+                    return None;
+                }
+
+                for (coord, weight) in self.coords_iter().zip(weights.iter().cycle()) {
+                    let (xi, yi) = coord.x_y();
+                    x_sum += xi * weight;
+                    y_sum += yi * weight;
+                    weight_sum += weight;
+                }
+
+                let x = x_sum / weight_sum;
+                let y = y_sum / weight_sum;
+
+                Some(Point::new(x, y))
+            }
+        }
+    };
+}
+
+impl_mean_center_option!(LineString);
```

Comment:
It looks like you're computing centroid here. 

Two things:

1. The blog uses one weight per geometry, not per coord in the geometry.

2. For linear features like a LineString, the blog post uses midpoint instead of centroid. Though I can't think of an application of this algorithm where using a midpoint would be more useful than a centroid. I know there are applications where you want the "middle" of the geometry to be within a geometry, which might not happen with centroid  (which is why people love polylabel), but in this case, we're eventually averaging a bunch of points together, so there's no guarantee that the output point will be inside of any geometry anyway.